### ISSUE TITLE
Wrong gradient value using `torch.distributed.nn.all_gather`

### ISSUE THREAD

**Issue by nguyentanthong** [pytorch/pytorch#120386]
https://github.com/pytorch/pytorch/blob/5a114f72bf0aa93268ba51707a20f99b79ed048d/torch/distributed/nn/functional.py#L330-L341

It looks like using `_Reduce_Scatter.apply` (or `_AlltoAll.apply`) to compute gradient generates wrong (random) gradient. We can fix this by using `dist.reduce_scatter` instead.

**Comment by nguyentanthong** [comment]
Here is a script to reproduce the bug:
```python
"""
torchrun --standalone --nproc_per_node=4  bug.py
"""
import os
import torch
import torch.distributed as dist
from torch.distributed import ReduceOp, init_process_group
from torch.distributed import nn as dist_nn

backend = "nccl"  # 'nccl', 'gloo', etc.
ddp = int(os.environ.get("RANK", -1)) != -1  # is this a ddp run?
if ddp:
    init_process_group(backend=backend)
    ddp_rank = int(os.environ["RANK"])
    ddp_local_rank = int(os.environ["LOCAL_RANK"])
    ddp_world_size = int(os.environ["WORLD_SIZE"])
    device = f"cuda:{ddp_local_rank}"
    torch.cuda.set_device(device)
    master_process = ddp_rank == 0  # this process will do logging, checkitemnting etc.
    seed_offset = ddp_rank  # each process gets a different seed
else:
    raise RuntimeError("DDP disabled")


class _AllGather(torch.autograd.Function):
    @staticmethod
    def forward(ctx, tensor):
        del ctx
        tensor = tensor.contiguous()
        out_tensor_list = [torch.empty_like(tensor) for _ in range(ddp_world_size)]
        dist.all_gather(out_tensor_list, tensor)
        return tuple(out_tensor_list)

    @staticmethod
    def backward(ctx, *grad_outputs):
        del ctx
        gx = torch.empty_like(grad_outputs[ddp_rank]).contiguous()
        input_tensor_list = tuple(t.contiguous() for t in grad_outputs)
        dist.reduce_scatter(gx, list(input_tensor_list), op=ReduceOp.SUM)
        return gx


all_gather = _AllGather.apply

def compute_grad(all_gather_fn):
    class AB(torch.nn.Module):
        def __init__(self):
            super().__init__()
            self.A = torch.nn.Linear(64, 128)
            self.B = torch.nn.Linear(64, 128)

        def forward(self, a, b):
            return self.A(a), self.B(b)

    torch.manual_seed(1337)
    net = AB().to(device=device)
    a = torch.randn(4, 16, 64).to(device=device)
    b = torch.randn(4, 16, 64).to(device=device)
    x, y = net(a[ddp_rank], b[ddp_rank])
    all_y = torch.cat(all_gather_fn(y), dim=0)
    logits = torch.einsum("AD,BD->AB", x, all_y)
    target = torch.arange(0, x.shape[0], 1, device=device, requires_grad=False) + ddp_rank * y.shape[0]
    loss = torch.nn.functional.cross_entropy(logits, target)
    loss.backward()
    g = net.B.weight.grad.sum()
    torch.distributed.all_reduce(g)
    return g.item()


for i in range(5):
    g = compute_grad(dist_nn.all_gather)
    if master_process:
        print( "Grad (pytorch)", g)

for i in range(5):
    g = compute_grad(all_gather)
    if master_process:
        print( "Grad (correct)", g)
```
